### PR TITLE
Python code refactoring.

### PR DESCRIPTION
**Pull request by Gadgetoid**
I got a little carried away, so this all ended up in one single commit. Raising as a draft until I can unpick it into some reasonable suggestions. Though feel free to cherry pick what matters to you.

It's also completely untested, but is thoroughly linted so I hopefully haven't done anything completely silly.

Some key changes:

* Rewrite terminal ID handling (picking up your todo) to use the stl urllib.parse, then split the path on "/" and grab the last element
* Move terminal ID creation into Terminal, and make it based on a UUID so it can't get potluck conflicting IDs (random has a tendency of not being random enough)
* Rewrite started, completed and alive to be a little more concise
* Try to use an explicit `encode("utf-8")` instead of a blunt convert to `bytes()`
* Try to make `get_terminal_env` a little cleaner
* Remove `Optional` from `cmdline` class variable, since `cmdline` is used regardless and could be undefined (raising an `AttributeError`)
* Rename from variables away from Python builtins (`min`, `id`) to avoid future mishaps

### CODE REVIEW

**Comment by Alex4386** [comment]
Will look into it! Thanks for the PR!

File: defaults/py_modules/decky_terminal/__init__.py [hunk 1]
```diff
@@ -26,36 +31,31 @@ def __init__(self) -> None:
...
+
     def _is_unix_shell_path(self, path: str) -> bool:
         return len(path) > 0 and path.startswith("/") and not path.isspace()
- 
+
     # CONFIG ================================================
     def get_config_filename(self) -> str:
-        return DECKY_PLUGIN_SETTINGS_DIR+os.sep+"config.json"
-    
+        return os.path.join(DECKY_PLUGIN_SETTINGS_DIR, "config.json")
```
Comment (Alex4386) [inline defaults/py_modules/decky_terminal/__init__.py:58]: Nice change 👍🏼

File: defaults/py_modules/decky_terminal/__init__.py [hunk 2]
```diff
@@ -81,107 +81,104 @@ async def get_default_shell(self) -> str:
...
 
     # TERMINAL CREATION =====================================
-    async def create_terminal(self, id: str, cmdline: Optional[str] = None):
+    async def create_terminal(self, terminal_id: str = None, cmdline: Optional[str] = None):
         if cmdline is None:
             cmdline = await self.get_default_shell()
 
-        print('cmdline!!!!', cmdline)
+        if terminal_id is None:
+            terminal_id = f"term-{uuid.uuid4()}"
```
Comment (Alex4386) [inline defaults/py_modules/decky_terminal/__init__.py]: The terminal ID can remove `term-` prefix since uuid itself is unique enough.  
The previous implementation had terminal_id of `term_####` format due to readability of uninitialized terminal having too long name on frontend. (the frontend fallbacks to terminal `id` when there is no `title` present)  

This can be mitigated via setting `title` of the terminal as `term_{last 4 digits of uuid}` or other `"user-identifiable measures"`

File: defaults/py_modules/decky_terminal/common.py [hunk 3]
```diff
@@ -30,18 +31,18 @@ async def write_file(cls, filename: str, content: str) -> bool:
...
-        
+
     @classmethod
     def merge_dict(cls, prev: Dict[_T, _U], new: Dict[_T, _U]) -> Dict[_T, _U]:
         for i, v in new.items():
             if prev.get(i) is None:
                 prev[i] = v
             else:
-                if type(prev.get(i)) is dict and type(v) is dict:
+                if isinstance(prev.get(i), dict) and isinstance(v, dict):
```
Comment (Alex4386) [inline defaults/py_modules/decky_terminal/common.py:43]: Nice cleanup 👍🏼

File: defaults/py_modules/decky_terminal/terminal.py [hunk 4]
```diff
@@ -105,37 +104,39 @@ async def _process_subscriber(self, ws: WebSocketServerProtocol):
...
     # PROCESS CONTROL =======================================
     def get_terminal_env(self):
-        result = dict(**os.environ)
-        result["TERM"] = "xterm-256color"
-        result["PWD"] = result["HOME"]
-        result["SSH_TTY"] = os.ttyname(self.slave_fd)
-        result["LINES"] = str(self.rows)
-        result["COLUMNS"] = str(self.cols)
-        result["XDG_RUNTIME_DIR"] = "/run/user/"+str(os.getuid())
+        result = dict(os.environ)  # dict(os.environ) == dict(**os.environ)
```
Comment (Alex4386) [inline defaults/py_modules/decky_terminal/terminal.py]: Nice cleanup 👍🏼

**Comment by Alex4386** [comment]
@Gadgetoid Looks OK! Since it is marked as WIP, Is it OK to merge into `main`?

**Review by Alex4386 — APPROVED** [review]

File: defaults/py_modules/decky_terminal/__init__.py (hunk 2 above)
Comment (Alex4386) [reply defaults/py_modules/decky_terminal/__init__.py]: Will implement that on frontend side after the merge. Thanks

Comment (Gadgetoid) [reply defaults/py_modules/decky_terminal/__init__.py]: Oh that's a good idea!

**Comment by Gadgetoid** [comment]
> @Gadgetoid Looks OK! Since it is marked as WIP, Is it OK to merge into `main`?

It remains *completely untested*, outside of some basic linting and code quality checks. 

If it works as you expect then it's probably okay to merge but you might want some more concise, individual commits (other than "Untested code savagery" in case any specific change breaks your code- since it'll be difficult to roll that one thing back amidst this mess. That's entirely up to you, though!

Might take me a while to come up with cleaner commits, though, since I'm buried in work and, wow, I really *did* get carried away here.

**Comment by Gadgetoid** [comment]
Just force pushed a, hopefully, verbose commit history. A diff of the changes versus my previous monolithic commit reveals the following differences.

TLDR: I've caught a couple of additional idiomatic python things, removed some logical redundancy and been a little more brutal removing dead code and let `black` have its way wrapping lines I don't necessarily agree with.

⚠️ Oh! And *still untested!*- I should probably rectify that.

```diff
diff --git a/defaults/py_modules/decky_terminal/__init__.py b/defaults/py_modules/decky_terminal/__init__.py
index 23b5916..f9199d8 100644
--- a/defaults/py_modules/decky_terminal/__init__.py
+++ b/defaults/py_modules/decky_terminal/__init__.py
@@ -44,13 +44,13 @@ class DeckyTerminal:
             if data is not None:
                 shells = data.splitlines()
                 shells = list(filter(self._is_unix_shell_path, shells))
-                if len(shells) > 0:
+                if shells:
                     return shells
 
             return ["/bin/sh"]
 
     def _is_unix_shell_path(self, path: str) -> bool:
-        return len(path) > 0 and path.startswith("/") and not path.isspace()
+        return path.startswith("/")
 
     # CONFIG ================================================
     def get_config_filename(self) -> str:
@@ -122,11 +122,11 @@ class DeckyTerminal:
         if cmdline is None:
             cmdline = await self.get_default_shell()
 
-        if terminal_id is None:
-            terminal_id = f"term-{uuid.uuid4()}"
-
         flags = await self._get_terminal_flags()
 
+        if terminal_id is None:
+            terminal_id = str(uuid.uuid4())
+
         if self._terminal_sessions.get(terminal_id) is None:
             self._terminal_sessions[terminal_id] = Terminal(cmdline, **flags)
 
diff --git a/defaults/py_modules/decky_terminal/terminal.py b/defaults/py_modules/decky_terminal/terminal.py
index 4f933cb..e60b4ea 100644
--- a/defaults/py_modules/decky_terminal/terminal.py
+++ b/defaults/py_modules/decky_terminal/terminal.py
@@ -35,7 +35,8 @@ class Terminal:
     _title_cache: bytes = b""
 
     def __init__(self, cmdline: str, is_shell: bool = True, **kwargs):
-        self.cmdline = cmdline  # cmdline can't meaningfully be None or undefined since it must be available for _start_process
+        self.cmdline = cmdline  # TODO: maybe raise ValueError? cmdline can't meaningfully be None or undefined since it must be available for _start_process
+
         self.is_shell = is_shell
         self.buffer = collections.deque([], maxlen=4096)
         self.subscribers = []
@@ -89,11 +90,15 @@ class Terminal:
 
         if self.master_fd is not None:
             new_size = struct.pack("HHHH", rows, cols, 0, 0)
-            await Common._run_async(fcntl.ioctl, self.master_fd, termios.TIOCSWINSZ, new_size)
+            await Common._run_async(
+                fcntl.ioctl, self.master_fd, termios.TIOCSWINSZ, new_size
+            )
 
         if self.slave_fd is not None:
             new_size = struct.pack("HHHH", rows, cols, 0, 0)
-            await Common._run_async(fcntl.ioctl, self.slave_fd, termios.TIOCSWINSZ, new_size)
+            await Common._run_async(
+                fcntl.ioctl, self.slave_fd, termios.TIOCSWINSZ, new_size
+            )
 
     # WORKERS ==============================================
     async def _process_subscriber(self, ws: WebSocketServerProtocol):
@@ -115,7 +120,7 @@ class Terminal:
 
     # PROCESS CONTROL =======================================
     def get_terminal_env(self):
-        result = dict(os.environ)  # dict(os.environ) == dict(**os.environ)
+        result = dict(os.environ)
 
         result.update({
             "TERM": "xterm-256color",
@@ -136,7 +141,6 @@ class Terminal:
 
     async def _start_process(self):
         self.master_fd, self.slave_fd = pty.openpty()
-        # self._set_pty_settings()
 
         await self._change_pty_size(self.rows, self.cols)
         self.process = await asyncio.create_subprocess_exec(
@@ -146,7 +150,6 @@ class Terminal:
             stderr=self.slave_fd,
             stdin=self.slave_fd,
             env=self.get_terminal_env(),
-            # creationflags=subprocess.CREATE_NO_WINDOW,
             cwd=os.getenv("HOME"),
         )
 
@@ -215,7 +218,9 @@ class Terminal:
         await Common._run_async(os.fsync, self.master_fd)
 
     async def _read_output(self) -> bytes:
-        output = await Common._run_async(os.read, self.master_fd, self._calculate_sync_size())
+        output = await Common._run_async(
+            os.read, self.master_fd, self._calculate_sync_size()
+        )
         if len(output) > 0:
             self._put_buffer(output)
             await self.broadcast_subscribers(output)
diff --git a/main.py b/main.py
index 6678be7..11942d1 100644
--- a/main.py
+++ b/main.py
@@ -1,12 +1,5 @@
 from typing import List, Optional
-
-import decky_plugin
-
-try:
-    from decky_terminal import DeckyTerminal
-except Exception as e:
-    decky_plugin.logger.info("[DeckyTerminal] Import Failed:", type(e), e)
-    raise e
+from decky_terminal import DeckyTerminal
 
 
 class Plugin:
@@ -24,24 +17,20 @@ class Plugin:
 
         for terminal_id, terminal in terminals.items():
             result = terminal.serialize()
-            output.append(
-                dict(
-                    id=terminal_id,
-                    **result,
-                )
-            )
+            result["id"] = terminal_id  # TODO: should a terminal own its ID?
+            output.append(result)
 
         return output
 
-    async def get_terminal(self, id: str) -> Optional[dict]:
-        terminal = Plugin.decky_terminal.get_terminal(id)
+    async def get_terminal(self, terminal_id: str) -> Optional[dict]:
+        terminal = Plugin.decky_terminal.get_terminal(terminal_id)
         if terminal is None:
             return None
 
         return terminal.serialize()
 
-    async def set_terminal_title(self, id: str, title: str) -> bool:
-        Plugin.decky_terminal.set_terminal_title(id, title)
+    async def set_terminal_title(self, terminal_id: str, title: str) -> bool:
+        Plugin.decky_terminal.set_terminal_title(terminal_id, title)
 
     async def create_terminal(self, terminal_id=None) -> bool:
         await Plugin.decky_terminal.create_terminal(terminal_id)
@@ -54,18 +43,15 @@ class Plugin:
         except:
             return False
 
-    async def change_terminal_window_size(
-        self, terminal_id, rows: int, cols: int
-    ) -> bool:
+    async def change_terminal_window_size(self, terminal_id, rows: int, cols: int) -> bool:
         try:
             terminal = Plugin.decky_terminal.get_terminal(terminal_id)
             if terminal is not None:
                 await terminal.change_window_size(rows, cols)
                 return True
+            return False
         except:
-            pass
-
-        return False
+            return False
 
     async def get_config(self) -> str:
         return await Plugin.decky_terminal.get_config()
@@ -90,10 +76,3 @@ class Plugin:
 
     async def _migration(self):
         pass
-
-    async def _internal_test(self):
-        await self._main()
-        decky_plugin.logger.info(Plugin.decky_terminal.get_server_port())
-
-        await Plugin.decky_terminal.create_terminal("testterm")
-        decky_plugin.logger.info("testterm created.")

```

**Comment by Gadgetoid** [comment]
Okay *now* I have done some cursory testing and quickly rebased out a bug in my urlparse code (`[:-1]` where I should have used `[-1]` ... that one always gets me).

**Comment by Gadgetoid** [comment]
Inspired by your idea for terminal naming, I wondered what it would look like if I converted the uuid to a phonetic alphabet. I chose nato arbitrarily, but you could get fun with it-

decky-terminal-screenshot

**Comment by Alex4386** [comment]
> Inspired by your idea for terminal naming, I wondered what it would look like if I converted the uuid to a phonetic alphabet. I chose nato arbitrarily, but you could get fun with it-
> 
> decky-terminal-screenshot
  
ICAO Phonetic Alphabet seems like a fun way of representing UUIDs! And also, thanks for testing. I was unable to test your branch due to other projects.  
If you think it's ready to merge, feel free to ping me!

**Comment by Alex4386** [comment]
> TLDR: I've caught a couple of additional idiomatic python things, removed some logical redundancy and been a little more brutal removing dead code and let `black` have its way wrapping lines I don't necessarily agree with.
> 

You've might have seen my code, or from my other projects, I'm not really a Python person.  
That was the reason the code was not pythonic. Thanks for rectifying that. 😆  
  
Maybe I should setup Blake or other linting action later on 🤔

**Comment by Gadgetoid** [comment]
I think the current in-vogue Python linting tool is “ruff”, I’ve been contemplating an upstream patch to the template to add that into CI since it can catch real actual bugs and not just syntax quirks.

“black” is a very opinionated reformatting tool that- depending on your preference- can either be useful, annoying or some mix of the two.

I’ll try and do some more in depth testing, currently the phonetic terminal titles don’t show up at the top of the individual terminals and I’m not a React person so I haven’t looked into that yet 🤣

I also had some ideas about making the Terminal class implicitly track open terminals/manage sessions, but that’s a little bit more ambitious and breaky!

**Comment by Alex4386** [comment]
OK, If you are ready. I'll merge it to the `main` branch.

**Comment by Alex4386** [comment]
Since this is no longer marked as WIP, I'll consider answer as Yes and merge.

**Comment by Gadgetoid** [comment]
Sorry got distracted. It was ready enough! I'm a prod away if testers catch anything that needs fixing, just punt it over.

**Comment by Alex4386** [comment]
Getting ready with v0.4.0 release: https://github.com/SteamDeckHomebrew/decky-plugin-database/pull/580

**Comment by Alex4386** [comment]
Merged and Now available on stores!
The Acknowledgement page has been updated to mark your work! See [Line 23 on AcknowledgementPage.tsx](https://github.com/Alex4386/decky-terminal/blob/main/src/pages/settings/AcknowledgementPage.tsx#L23).

Thank you for your work! 😊 @Gadgetoid
